src/utilities/util_dump.c

```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
+    if (fprintf(fp, "%s%s%s%s%s", prefix, current_key, infix, value, suffix) < 0)
+        return (util_err(session, EIO, NULL));
+    return (0);
+}
 
-        } else {
-            /* Continue parsing the file. */
-            ret = reverse ? cursor->prev(cursor) : cursor->next(cursor);
-            if (ret == 0) {
-                if ((ret = cursor->get_key(cursor, &current_key)) != 0)
-                    return (util_cerr(cursor, "get_key", ret));
-                if ((ret = cursor->get_value(cursor, &value)) != 0)
-                    return (util_cerr(cursor, "get_value", ret));
-            } else if (ret != WT_NOTFOUND)
-                return (util_cerr(cursor, (reverse ? "prev" : "next"), ret));
+/*
+ * dump_record --
+ *     Dump the record specified by key or one near to it. If a window is specified print out up to
+ *     that many records before and after sought record. The window will be truncated if it would
+ *     move past the first or last entry.
+ */
+static int
+dump_record(
+  WT_CURSOR *cursor, const char *key, bool reverse, bool search_near, bool json, uint64_t window)
+{
+    WT_DECL_RET;
+    WT_SESSION *session;
+    int exact;
+    const char *current_key;
+    bool once;
+    unsigned int n, total_window;
+    int (*fwd)(WT_CURSOR *);
+    int (*bck)(WT_CURSOR *);
+
+    session = cursor->session;
+    once = false;
+    exact = 0;
+
+    WT_ASSERT((WT_SESSION_IMPL *)session, key != NULL);
+
+    current_key = key;
+    cursor->set_key(cursor, current_key);
+    ret = cursor->search_near(cursor, &exact);
+
+    if (ret != 0 && ret != WT_NOTFOUND)
+        return (util_cerr(cursor, "search_near", ret));
+
+    /*
+     * If a key has been found but search near is disabled and there is no exact match, the
+     * requested key is missing.
+     */
+    if (ret == 0 && !search_near && exact != 0)
+        return (WT_NOTFOUND);
+
+    if (window <= 0)
+        ret = print_record(cursor, json);
+    else {
+        fwd = (reverse) ? cursor->prev : cursor->next;
+        bck = (reverse) ? cursor->next : cursor->prev;
+
+        /* Back up as far as possible in the window. */
+        for (n = 0; n <= window; n++) {
+            if ((ret = bck(cursor)) != 0) {
+                if (ret == WT_NOTFOUND)
+                    break;
+                return (util_cerr(cursor, "cursor", ret));
+            }
         }
 
-        if (ret == 0) {
-            if (fprintf(fp, "%s%s%s%s%s%s", json && once ? "," : "", prefix, current_key, infix,
-                  value, suffix) < 0)
-                return (util_err(session, EIO, NULL));
+        /* Calculate the maximum window size based on how far it was possible to back up in the
+         * window above.
+         *
+         * If it was not possible to backup (n==0) it is then necessary to include the searched for
+         * record in the total window size.
+         */
+        total_window = ((n == 0) ? 1 : n) + window;
+
+        for (n = 0; n < total_window; n++) {
```

Comment:
How are we preventing ourselves from dumping the key/value pairs twice if we are walking from point we walked back to?

Reply:
Ah I misread. Gotcha.